smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/UserTuneManager.java

```diff
@@ -0,0 +1,167 @@
+                            listener.onUserTuneUpdated(contact, message, tune);
+                        }
+                    }
+                });
+            }
+        });
+    }
+
+    public static synchronized UserTuneManager getInstanceFor(XMPPConnection connection) {
+        UserTuneManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new UserTuneManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    public void setUserTune(UserTune userTune) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        setUserTune(userTune,null,null);
+    }
+
+    public void setUserTune(UserTune userTune,String text) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        setUserTune(userTune, null, text);
+    }
+
+    public void setUserTune(UserTune userTune,UserTuneConcretisation concretisation) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        setUserTune(userTune, concretisation, null);
+    }
+
+    public void setUserTune(UserTune userTune,UserTuneConcretisation concretisation,String text) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        UserTuneElement element = buildUserTune(userTune,concretisation,text);
+        publishUserTune(element);
+    }
+
+    public void clearUserTune() throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        UserTuneElement element = buildUserTune(null,null,null);
+        publishUserTune(element);
+    }
+
+    private void publishUserTune(UserTuneElement userTuneElement) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        if (pubSubManager == null) {
+            pubSubManager = PubSubManager.getInstance(getAuthenticatedConnectionOrThrow(), connection().getUser().asBareJid());
+        }
+        LeafNode node = pubSubManager.getOrCreateLeafNode(USERTUNE_NODE);
+        node.publish(new PayloadItem<>(userTuneElement));
+    }
+
+    private static UserTuneElement buildUserTune(UserTune userTune,UserTuneConcretisation concretisation,String text) {
+        return new UserTuneElement(
+                new UserTuneElement.UserTuneSubjectElement(userTune,concretisation));
+    }
+
+    public static void addUserTuneToMessage(Message message,UserTune userTune) {
+        addUserTuneToMessage(message, userTune, null);
+    }
+
+    public static void addUserTuneToMessage(Message message,UserTune userTune,UserTuneConcretisation concretisation) {
```

Comment:
You might want to move methods like `addUserTuneToMessage()` into the UserTuneElement class and reference those methods from here. That way you can also create a method like `getUserTuneElementFromMessage` inside the `UserTuneElement` class and have both utility methods for adding and retrieving elements from a message in one place.

Reply:
I see you already did this further below :+1: